kubernetes/kubernetes_service.go

```diff
@@ -236,3 +267,21 @@ func (in *IstioClient) getDeployments(namespace string, deploymentsChan chan dep
 	deployments, err := in.k8s.AppsV1beta1().Deployments(namespace).List(emptyListOptions)
 	deploymentsChan <- deploymentsResponse{deployments: deployments, err: err}
 }
+
+func (in *IstioClient) getDeployment(namespace string, deploymentName string) (*v1beta1.Deployment, error) {
+	return in.k8s.AppsV1beta1().Deployments(namespace).Get(deploymentName, emptyGetOptions)
+}
+
+func (in *IstioClient) getPods(namespace string, selector map[string]string, podsChan chan podsResponse) {
+	selectorQuery := labels.Set(selector).String()
+	pods, err := in.GetPods(namespace, selectorQuery)
+	podsChan <- podsResponse{pods: pods, err: err}
```

Comment:
reviewers, do you mind confirming that I am fetching the correct pods from a deployment?

Reply:
I am not 100% sure if it's ok, because above you're passing `deployment.Spec.Selector.MatchLabels` as the selector, thus ignoring `deployment.Spec.Selector.MatchExpressions`, and I can't tell if it's fine or not (that is, I don't know if the information in `MatchExpressions` is redundant with `MatchLabels` or nor).

Sometimes I wonder if we're not making everything over-complicated with all the labels stuff and we should not just get the whole list of pods in namespace and look for their ref owner. But that would be a bit less efficient to query all the pods. @lucasponce WDYT?

Reply:
This impact in https://github.com/kiali/kiali/pull/393 too. I need to get the pods for check the istioSidecar and I am using Matchlabels like Selector ignoring MatchExpressions

```go
type LabelSelector struct {
	// matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
	// map is equivalent to an element of matchExpressions, whose key field is "key", the
	// operator is "In", and the values array contains only "value". The requirements are ANDed.
	// +optional
	MatchLabels map[string]string `json:"matchLabels,omitempty" protobuf:"bytes,1,rep,name=matchLabels"`
	// matchExpressions is a list of label selector requirements. The requirements are ANDed.
	// +optional
	MatchExpressions []LabelSelectorRequirement `json:"matchExpressions,omitempty" protobuf:"bytes,2,rep,name=matchExpressions"`
}
```

I think that we need to use MatchExpressions too. It's like is used to validate for example if version is equal to one of [ 'v1','v2']

```go
// A label selector requirement is a selector that contains values, a key, and an operator that
// relates the key and values.
type LabelSelectorRequirement struct {
	// key is the label key that the selector applies to.
	// +patchMergeKey=key
	// +patchStrategy=merge
	Key string `json:"key" patchStrategy:"merge" patchMergeKey:"key" protobuf:"bytes,1,opt,name=key"`
	// operator represents a key's relationship to a set of values.
	// Valid operators are In, NotIn, Exists and DoesNotExist.
	Operator LabelSelectorOperator `json:"operator" protobuf:"bytes,2,opt,name=operator,casttype=LabelSelectorOperator"`
	// values is an array of string values. If the operator is In or NotIn,
	// the values array must be non-empty. If the operator is Exists or DoesNotExist,
	// the values array must be empty. This array is replaced during a strategic
	// merge patch.
	// +optional
	Values []string `json:"values,omitempty" protobuf:"bytes,3,rep,name=values"`
}
```

Reply:
There is a helper that convert the labelSelector (MatchLabels & MatchExpressions ) to map[string]string we can use this and then pass this like a selector in the filter.

```go
// LabelSelectorAsMap converts the LabelSelector api type into a map of strings, ie. the
// original structure of a label selector. Operators that cannot be converted into plain
// labels (Exists, DoesNotExist, NotIn, and In with more than one value) will result in
// an error.
func LabelSelectorAsMap(ps *LabelSelector) (map[string]string, error) {
	if ps == nil {
		return nil, nil
	}
	selector := map[string]string{}
	for k, v := range ps.MatchLabels {
		selector[k] = v
	}
	for _, expr := range ps.MatchExpressions {
		switch expr.Operator {
		case LabelSelectorOpIn:
			if len(expr.Values) != 1 {
				return selector, fmt.Errorf("operator %q without a single value cannot be converted into the old label selector format", expr.Operator)
			}
			// Should we do anything in case this will override a previous key-value pair?
			selector[expr.Key] = expr.Values[0]
		case LabelSelectorOpNotIn, LabelSelectorOpExists, LabelSelectorOpDoesNotExist:
			return selector, fmt.Errorf("operator %q cannot be converted into the old label selector format", expr.Operator)
		default:
			return selector, fmt.Errorf("%q is not a valid selector operator", expr.Operator)
		}
	}
	return selector, nil
}

```